### ISSUE TITLE
NDVI hybrid green correction triggers early dask computations

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#2619]
**Describe the bug**

Loading the `true_color` composite for FCI triggers early dask computations, affecting performance negatively. This happens due to the NDVI hybrid green correction calling `da.where` with an `xarray.DataArray`.

**To Reproduce**

```python
# Your code here
import hdf5plugin
from satpy import Scene
from glob import glob
from satpy.utils import debug_on; debug_on()
from satpy.tests.utils import CustomScheduler
import dask.config
fci_files = glob("/media/nas/x21308/MTG_test_data/2022_05_MTG_Testdata/RC0040/*BODY*.nc")
with dask.config.set(scheduler=CustomScheduler(max_computes=1)):
    sc = Scene(filenames={"fci_l1c_nc": fci_files})
    sc.load(["true_color"])
    sc.save_datasets()
```

**Expected behavior**

I expect this to complete without errors.

**Actual results**

Limiting to the final part for brevity:

```
[INFO: 2023-10-26 12:16:42 : satpy.writers] Computing and writing results...
Traceback (most recent call last):
  File "/data/gholl/checkouts/protocode/mwe/satpy-fci-too-many-computes.py", line 11, in <module>
    sc.save_datasets()
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/scene.py", line 1283, in save_datasets
    return writer.save_datasets(dataarrays, compute=compute, **save_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/writers/__init__.py", line 756, in save_datasets
    return compute_writer_results([results])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/writers/__init__.py", line 594, in compute_writer_results
    da.compute(delayeds)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/base.py", line 628, in compute
    results = schedule(dsk, keys, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/tests/utils.py", line 288, in __call__
    raise RuntimeError("Too many dask computations were scheduled: "
RuntimeError: Too many dask computations were scheduled: 4
```

**Screenshots**

**Environment Info:**
 - OS: openSUSE 15.3
 - Satpy Version: main
 - 
**Additional context**

This is captured only at the 4th rather than the 2nd computation due to #2615.

Setting a breakpoint in `tests.utils:288` reveals the early computation happens due to a `da.where` call during the NDVI hybrid green correction:

```python-traceback
  File "/data/gholl/checkouts/protocode/mwe/satpy-fci-too-many-computes.py", line 10, in <module>
    sc.load(["true_color"])
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/scene.py", line 1445, in load
    self.generate_possible_composites(unload)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/scene.py", line 1508, in generate_possible_composites
    keepables = self._generate_composites_from_loaded_datasets()
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/scene.py", line 1527, in _generate_composites_from_loaded_datasets
    return self._generate_composites_nodes_from_loaded_datasets(needed_comp_nodes)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/scene.py", line 1533, in _generate_composites_nodes_from_loaded_datasets
    self._generate_composite(node, keepables)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/scene.py", line 1591, in _generate_composite
    composite = compositor(prereq_datasets,
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/composites/spectral.py", line 169, in __call__
    ndvi.data = da.where(ndvi > self.ndvi_min, ndvi, self.ndvi_min)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/array/routines.py", line 2107, in where
    return elemwise(np.where, condition, x, y)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/array/core.py", line 4831, in elemwise
    result = blockwise(
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/array/blockwise.py", line 286, in blockwise
    meta = compute_meta(func, dtype, *args[::2], **kwargs)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/array/utils.py", line 140, in compute_meta
    meta = func(*args_meta, **kwargs_meta)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/xarray/core/common.py", line 165, in __array__
    return np.asarray(self.values, dtype=dtype)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/xarray/core/dataarray.py", line 761, in values
    return self.variable.values
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/xarray/core/variable.py", line 555, in values
    return _as_array_or_item(self._data)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/xarray/core/variable.py", line 305, in _as_array_or_item
    data = np.asarray(data)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/array/core.py", line 1700, in __array__
    x = self.compute()
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/base.py", line 342, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/dask/base.py", line 628, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/satpy/tests/utils.py", line 288, in __call__
    raise RuntimeError("Too many dask computations were scheduled: "
```

inspection in `composites.spectral.py` reveals that `da.where` is given `xarray.DataArray` types which then triggers a `compute_meta` calculation, just like in #2614 for the day/night composite.

Apart from the day/night-composite and the NDVI hybrid green correction, there might be other places in Satpy where we are passing `xarray.DataArray` to `da.where`.

**Comment by gerritholl** [comment]
@strandgren Any thoughts?

**Comment by strandgren** [comment]
sorry about that, a bad side effect of my limited understanding of dask.. If you have some suggestion on how to fix it I'm happy to make a PR to fix it.

**Comment by strandgren** [comment]
Or does it make more sense to make a PR that fixes all known cases of this (including e.g. #2614 as well)?

**Comment by pnuu** [comment]
I wrapped the compositor creation and calls of `TestNdviHybridGreenCompositor` cases with the custom scheduler and don't see any computes.

**Comment by gerritholl** [comment]
> Or does it make more sense to make a PR that fixes all known cases of this (including e.g. #2614 as well)?

#2614 was already fixed in #2617, but I suspect that if we get a fix for #2615 and https://github.com/dask/dask/issues/10595 that we might discover many more computes that so far went unnoticed…

**Comment by pnuu** [comment]
I'll have a go fixing this while waiting for FCI data.
